src/codec.rs

```diff
@@ -0,0 +1,474 @@
+//! small number of specializations for u8, u16 and maybe u24, we do it by hand.
+
+use std::{
+    convert::TryInto,
+    io::{Cursor, Read},
+    mem::size_of,
+};
+
+/// Describes how to decode an object from a byte sequence.
+pub trait Decode<D>: Sized {
+    /// Type of errors returned by the [`decode`] method.
+    type Error;
+
+    /// Read and decode an encoded object from `bytes`. `decoding_parameter` provides details of the
+    /// wire encoding such as lengths of different portions of the message.
+    fn decode(decoding_parameter: &D, bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error>;
+
+    /// Convenience method to get decoded value.
+    fn get_decoded(decoding_parameter: &D, bytes: &[u8]) -> Result<Self, Self::Error> {
+        let mut cursor = Cursor::new(bytes);
+        Self::decode(decoding_parameter, &mut cursor)
+    }
+}
+
+/// Describes how to encode objects into a byte sequence.
+pub trait Encode {
+    /// Append the encoded form of this object to the end of `bytes`, growing the vector as needed.
+    fn encode(&self, bytes: &mut Vec<u8>);
+
+    /// Convenience method to get encoded value.
+    fn get_encoded(&self) -> Vec<u8> {
+        let mut ret = Vec::new();
+        self.encode(&mut ret);
+        ret
+    }
+}
+
+impl Decode<()> for u8 {
+    type Error = std::io::Error;
+
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error> {
+        let mut value = [0u8; size_of::<u8>()];
+        bytes.read_exact(&mut value)?;
+        Ok(value[0])
+    }
+}
+
+impl Encode for u8 {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.push(*self);
+    }
+}
+
+impl Decode<()> for u16 {
+    type Error = std::io::Error;
+
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error> {
+        let mut value = [0u8; size_of::<u16>()];
+        bytes.read_exact(&mut value)?;
+        Ok(u16::from_be_bytes(value))
+    }
+}
+
+impl Encode for u16 {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend_from_slice(&u16::to_be_bytes(*self));
+    }
+}
+
+impl Decode<()> for u64 {
+    type Error = std::io::Error;
+
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, Self::Error> {
+        let mut value = [0u8; size_of::<u64>()];
+        bytes.read_exact(&mut value)?;
+        Ok(u64::from_be_bytes(value))
+    }
+}
+
+impl Encode for u64 {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend_from_slice(&u64::to_be_bytes(*self));
+    }
+}
+
+/// Encode `items` into `bytes` as a variable-length opaque vector with a
+/// maximum length of `0xff`
+pub fn encode_items_opaque_u8<E: Encode>(bytes: &mut Vec<u8>, items: &[E]) {
+    let len_offset = bytes.len();
+    bytes.push(0);
+
+    for i in items {
+        i.encode(bytes);
+    }
+
+    let len = bytes.len() - len_offset - 1;
+    debug_assert!(len <= 0xff);
```

Comment:
Like I said, I don't object to taking @BranLwyd's suggestion. I just wanted to point out that the current pattern is not atypical. (On the other hand, I've also seen libraries that return an error rather than panic. Doing so is not very ergonomic in Go or C++, but it's definitely ergonomic in Rust :) )